src/preprocessing/passes/learned_rewrite.cpp

```diff
@@ -155,14 +155,293 @@ Node LearnedRewrite::rewriteLearnedRec(Node n,
                                        std::unordered_set<Node>& lems,
                                        std::unordered_map<TNode, Node>& visited)
 {
-  return n;
+  NodeManager* nm = NodeManager::currentNM();
+  std::unordered_map<TNode, Node>::iterator it;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    it = visited.find(cur);
+
+    if (it == visited.end())
+    {
+      visited[cur] = Node::null();
```

Comment:
Is this completely standard in your preprocessing passes? If not (and maybe even if yes), please consider adding a comment about the meaning of not being in the map vs. being assigned null node in the map vs. being assigned a non-null node in the map.

Reply:
This is very common pattern in all "node conversion" methods, we store "null" to mark "pre-visited but not post-visited".